spython/main/parse/parsers/docker.py

```diff
@@ -9,25 +9,57 @@
 import os
 import re
 
-from .environment import parse_env
-from .recipe import Recipe
 from spython.logger import bot
+from .base import ParserBase
 
-class DockerRecipe(Recipe):
 
-    def __init__(self, recipe=None):
-        '''a Docker recipe parses a Docker Recipe into the expected fields of
-           labels, environment, and install/runtime commands. We save working
-           directory as we parse, and the last one can be added to the runscript
-           of a Singularity recipe.
+class DockerParser(ParserBase):
+
+    name = 'docker'
+
+    def __init__(self, recipe='Dockerfile', load=True):
+        '''a docker parser will read in a Dockerfile and put it into a Recipe
+           object.
 
            Parameters
            ==========
-           recipe: the recipe file (Dockerfile) to parse
+           recipe: the Dockerfile to parse. If not defined, deafults to 
+                   Dockerfile assumed to be in the $PWD.
 
         '''
-        self.name = "docker"
-        super(DockerRecipe, self).__init__(recipe)
+        super(DockerParser, self).__init__(recipe, load)
+
+
+    def parse(self):
+        '''parse is the base function for parsing the Dockerfile, and extracting
+           elements into the correct data structures. Everything is parsed into
+           lists or dictionaries that can be assembled again on demand. 
+
+           Environment: Since Docker also exports environment as we go, 
+                        we add environment to the environment section and 
+                        install
+
+           Labels: include anything that is a LABEL, ARG, or (deprecated)
+                   maintainer.
+
+           Add/Copy: are treated the same
+
+        '''
+        parser = None
+        previous = None
+
+        for line in self.lines:
+
+            parser = self._get_mapping(line, parser, previous)
+
+            # Parse it, if appropriate
+            if parser:
+                parser(line)
+
+            previous = line
+
+        # Instantiated by ParserBase
```

Comment:
Actually: Why? Just do it here. No comments are better than required comments

Reply:
Because the base handles setting the filename, ensuring that it exists, and loading it into the recipe:

```
        self.filename = filename
        self._run_checks()
        self.recipe = Recipe(self.filename)
```
This makes it easy for any new kind of recipe parser to be written - the writer doesn't need to worry about this. This is a design decision that I've made and I think is good.

Reply:
And why not just move the line ` self.recipe = Recipe(self.filename)` here and set it to `None` in the base class? Also filename validation is not really good that that point because filename could still be `None` when you reach this point (actual parsing). So rather do something like: `validateFilepath(self.filename); self.recipe = Recipe(self.filename)` here